### PR TITLE
Smartcard tighter input validation

### PR DESCRIPTION
**Pull request by akallabeth**
* Unified NDR pointer reading when unpacking smartcard data from the RDP channel.
* Added a simple test program to trigger (most) API calls from within a FreeRDP session
* Unified NDR writing for responses

### CODE REVIEW

**Comment by akallabeth** [comment]
@bmiklautz @ondrejholy I´ve refactored the smartcard data decoding quite a bit and added a simple test program to have a small test case for each API call.
Fixed a few issues in previous decoding, but I´d feel better if a few more eyes could have a look at this.

File: channels/smartcard/client/smartcard_operations.c [hunk 1]
```diff
@@ -9,6 +9,8 @@
  * Copyright 2015 DI (FH) Martin Haimberger <martin.haimberger@thincast.com>
  * Copyright 2017 Armin Novak <armin.novak@thincast.com>
  * Copyright 2017 Thincast Technologies GmbH
+ * Copyright 2020 Armin Novak <armin.novak@thincast.com>
```
Comment (hardening) [inline channels/smartcard/client/smartcard_operations.c]: Duplicated copyright claim (present some line above)

File: channels/smartcard/client/smartcard_pack.c [hunk 2]
```diff
@@ -79,38 +178,149 @@ static LONG smartcard_ndr_read(wStream* s, BYTE** data, size_t min)
...
 	return STATUS_SUCCESS;
 }
 
-static LONG smartcard_ndr_read_a(wStream* s, CHAR** data, size_t min)
+static BOOL smartcard_ndr_pointer_write(wStream* s, UINT32* index, DWORD length)
+{
+	const UINT32 ndrPtr = 0x20000 + (*index) * 4;
+
+	if (!s)
+		return FALSE;
```
Comment (hardening) [inline channels/smartcard/client/smartcard_pack.c]: unneeded check

File: channels/smartcard/client/smartcard_operations.c (hunk 1 above)
Comment (akallabeth) [reply channels/smartcard/client/smartcard_operations.c]: oops ^^

**Comment by akallabeth** [comment]
@hardening tried to keep it small but during testing one bug after the next appeared :/

**Comment by fbezdeka** [comment]
Still no reviewer available? I would love to see the smartcard forwarding working again.

**Comment by ondrejholy** [comment]
One of our customers replied that this MR works for them only with /client-build-number:2600. I don't have more info (apart from that it doesn't solve https://github.com/FreeRDP/FreeRDP/issues/5792 for them, which I suppose is not expected with such build number). I've asked them to collect debug output. I will let you know once I will receive it.

File: channels/smartcard/client/smartcard_pack.c [hunk 3]
```diff
@@ -5,6 +5,8 @@
  * Copyright 2014 Marc-Andre Moreau <marcandre.moreau@gmail.com>
  * Copyright 2015 Thincast Technologies GmbH
  * Copyright 2015 DI (FH) Martin Haimberger <martin.haimberger@thincast.com>
+ * Copyright 2020 Armin Novak <armin.novak@thincast.com>
+ * Copyright 2020 Thincast Technologies GmbH
  *
```
Comment (hardening) [inline channels/smartcard/client/smartcard_pack.c]: duplicated copyright statement

File: channels/smartcard/client/smartcard_pack.c [hunk 4]
```diff
@@ -79,38 +178,149 @@ static LONG smartcard_ndr_read(wStream* s, BYTE** data, size_t min)
...
+
+static LONG smartcard_ndr_write_state(wStream* s, const ReaderState_Return* data, UINT32 size,
+                                      ndr_ptr_t type)
+{
+	union {
+		const ReaderState_Return* reader;
+		const BYTE* data;
+	} cnv;
+
+	cnv.reader = data;
```
Comment (hardening) [inline channels/smartcard/client/smartcard_pack.c]: @akallabeth I don't see in these 3 functions why we need a union, can you elaborate ?

**Comment by hardening** [comment]
General remark, there's a lot of statements like
```C
status = op(); 
if (status) 
   return status;
``` 
that assume that `SCARD_S_SUCCESS` is 0. Shouldn't we change them with
```C
status = op(); 
if (status != SCARD_S_SUCCESS) 
   return status;
```
?

File: channels/smartcard/client/smartcard_pack.h [hunk 5]
```diff
@@ -5,6 +5,8 @@
  * Copyright 2014 Marc-Andre Moreau <marcandre.moreau@gmail.com>
  * Copyright 2015 Thincast Technologies GmbH
  * Copyright 2015 DI (FH) Martin Haimberger <martin.haimberger@thincast.com>
+ * Copyright 2020 Armin Novak <armin.novak@thincast.com>
+ * Copyright 2020 Thincast Technologies GmbH
```
Comment (hardening) [inline channels/smartcard/client/smartcard_pack.h]: duplicated copyright statement

File: scripts/test-scard.cpp [hunk 6]
```diff
@@ -0,0 +1,921 @@
+/**
+ * FreeRDP: A Remote Desktop Protocol Implementation
+ * Smartcard API test program
+ *
+ * This simple program can be used to trigger calls for (almose) the
+ * entire SCARD API.
```
Comment (hardening) [inline scripts/test-scard.cpp]: s/almose/almost/

File: channels/smartcard/client/smartcard_pack.c (hunk 4 above)
Comment (akallabeth) [reply channels/smartcard/client/smartcard_pack.c]: this is the only way in C to cast away const ;)

**Comment by akklaw** [comment]
This patch works well except I need to specify /client-build-number:2000  as (XP) in order to get the smartcard redirection to work properly.  The version in current master as of 3/13/2020 does not need to specify this option but it works.

**Comment by akallabeth** [comment]
@akklaw I´ve added a 'hack' that was removed in this pr in https://github.com/akallabeth/FreeRDP/tree/smartcard_ndr_strict_ext - could you try if that fixes your issue with the status calls?

**Comment by fbezdeka** [comment]
Connecting to a Windows 10 system works when starting a new session. 

```
[15:26:44:264] [47610:47611] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[15:26:44:264] [47610:47611] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[15:26:44:288] [47610:47611] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[15:26:44:289] [47610:47620] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)
[15:26:44:289] [47610:47622] [INFO][com.freerdp.channels.rdpsnd.client] - Loaded fake backend for rdpsnd
[15:26:44:640] [47610:47620] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
[15:26:50:195] [47610:47611] [INFO][com.freerdp.client.x11] - Logon Error Info SESSION_ID [LOGON_MSG_SESSION_CONTINUE]
[15:26:50:351] [47610:47620] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
[15:26:55:493] [47610:47672] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardControl (0x000900D4), status: SCARD_E_SHARING_VIOLATION (0x8010000B)
[15:26:58:663] [47610:47630] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:27:01:430] [47610:47638] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:27:49:772] [47610:47784] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:28:15:469] [47610:47611] [INFO][com.freerdp.core] - ERRINFO_RPC_INITIATED_DISCONNECT_BY_USER (0x0000000B):The disconnection was initiated by an administrative tool on the server running in the user's session.
[15:28:15:622] [47610:47684] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:28:15:622] [47610:47621] [ERROR][com.freerdp.channels.rdpdr.client] - pVirtualChannelWriteEx failed with CHANNEL_RC_BAD_CHANNEL_HANDLE [00000007]
[15:28:15:622] [47610:47621] [ERROR][com.freerdp.channels.smartcard.client] - smartcard_complete_irp failed with error 7!
[15:28:15:640] [47610:47708] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
```

When trying to resume to an existing session, the smartcard redirection fails. 
System tray icon of my smartcard driver says "Service not started":

```
[15:28:35:749] [48443:48444] [INFO][com.freerdp.gdi] - Local framebuffer format  PIXEL_FORMAT_BGRX32
[15:28:35:749] [48443:48444] [INFO][com.freerdp.gdi] - Remote framebuffer format PIXEL_FORMAT_RGB16
[15:28:35:785] [48443:48444] [INFO][com.winpr.clipboard] - initialized POSIX local file subsystem
[15:28:35:785] [48443:48453] [INFO][com.freerdp.channels.rdpdr.client] - Loading device service smartcard (static)
[15:28:35:786] [48443:48455] [INFO][com.freerdp.channels.rdpsnd.client] - Loaded fake backend for rdpsnd
[15:28:35:027] [48443:48453] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
[15:28:39:183] [48443:48467] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:28:39:205] [48443:48471] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetStatusChangeW (0x000900A4), status: SCARD_E_CANCELLED (0x80100002)
[15:28:39:274] [48443:48505] [ERROR][com.freerdp.channels.smartcard.client] - SCardBeginTransaction failed with error -2146435069
[15:28:39:274] [48443:48505] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardBeginTransaction (0x000900BC), status: SCARD_E_INVALID_HANDLE (0x80100003)
[15:28:39:349] [48443:48453] [INFO][com.freerdp.channels.rdpdr.client] - registered device #1: SCARD (type=32 id=1)
```

**Comment by akklaw** [comment]
@akallabeth sync and running the latest changes, it removed that SCardGetStatusChangeW error but it doesn't detect the smartcard.  I tried client-build-number:2000, :20000, :4500, they are all failed to detect the smartcard.

**Comment by akallabeth** [comment]
@fbezdeka looks like a local `pcscd` issue, could you check that?

**Comment by ondrejholy** [comment]
> One of our customers replied that this MR works for them only with /client-build-number:2600. I don't have more info (apart from that it doesn't solve #5792 for them, which I suppose is not expected with such build number). I've asked them to collect debug output. I will let you know once I will receive it.

They see the already reported `SCARD_E_SHARING_VIOLATION` and `SCARD_E_CANCELLED` errors when they do not specify the `/client-build-number`. Apart from that, they also see the following `SCARD_E_UNSUPPORTED_FEATURE` errors:

```
[20:12:35:904] [32368:32383] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetDeviceTypeIdW failed with error SCARD_E_UNSUPPORTED_FEATURE [-2146435038]
[20:12:35:904] [32368:32383] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetDeviceTypeId (0x00090108), status: SCARD_E_UNSUPPORTED_FEATURE (0x80100022)
...
[20:12:38:110] [32368:32409] [ERROR][com.freerdp.channels.smartcard.client] - SCardGetReaderIconW failed with error SCARD_E_UNSUPPORTED_FEATURE [-2146435038]
[20:12:38:110] [32368:32409] [WARN][com.freerdp.channels.smartcard.client] - IRP failure: SCardGetReaderIcon (0x00090104), status: SCARD_E_UNSUPPORTED_FEATURE (0x80100022)
```

It doesn't work for them either with `/client-build-version:7601` and the log contains the already reported `SCARD_E_CANCELLED` errors.

**Comment by fbezdeka** [comment]
When building your branch `/client-build-version` seems to be an unsupported parameter. Is there some other branch required to get this? Maybe a merge with current master? (Sorry I did not check the history there...)

@ondrejholy `pcscd` does not report any error. So I guess everything is OK. It does not help to reboot the local system and try to join an already open Windows session afterwards. Logging out from Windows first and a following reconnect is the only scenario that I got working.

**Comment by akallabeth** [comment]
@ondrejholy the unsupported feature is expected (pcsclite can't handle the calls and tells the server)
As for the canceled stuff, that could get interesting as this seems to change how the device status call is used.

**Comment by akallabeth** [comment]
@fbezdeka sorry, typo, `client-build-number` not version

**Comment by fbezdeka** [comment]
This is a log when trying to connect to a Windows 10 system without specifying any special `/client-build-number:`. I'm connected to the destination system using a VPN connection. Just in case the "SCARD_E_TIMEOUT" errors may be related to that.

[xfreerdp-no-client-build-number.log](https://github.com/FreeRDP/FreeRDP/files/4397008/xfreerdp-no-client-build-number.log)

Result: The forwarding itself seems to work. Card is detected. When trying to use the card (windows application requires SmartCard auth) I'm asked for which smart card device to use. First there is a "Checking card status" phase, then it switches to "Ready to use". After choosing the presented smartcard device I'm never asked for a PIN. The application just reports an error.


When using `/client-build-number:2000` the situation changes a bit. The "choose which smartcard to use" dialog does not show up. I'm asked for my PIN instead. The first try to enter the PIN always fails. The second try succeeds.

[xfreerdp-client-build-number-2000.log](https://github.com/FreeRDP/FreeRDP/files/4397025/xfreerdp-client-build-number-2000.log)

**Comment by akallabeth** [comment]
@fbezdeka thank you, looks like there is an error with the transmit count call.

**Comment by akallabeth** [comment]
@fbezdeka dou you use windows default tools to test? Would be great if I could reproduce with the same tool/options to fix this.

**Comment by fbezdeka** [comment]
On the target system there is some smartcard "middleware" or "driver" named "CardOS API" installed. Don't know if this really matters, because it works with current master (at least when opening a new Windows session). The "server" or remote system is a Windows 10 1809.

**Comment by akallabeth** [comment]
@fbezdeka the latest commit should fix the transmit count IRP errors.

**Comment by hardening** [comment]
@mfleisz @bmiklautz any opportunity to look at this (big) PR ?

**Comment by fbezdeka** [comment]
@akallabeth thanks for trying to fix my problem.
I retried with 2749de5 but did not notice any difference. 

[FreeRDP-2749de506c.log](https://github.com/FreeRDP/FreeRDP/files/4407071/FreeRDP-2749de506c.log)

Would be very kind if you could have another look into the log. Thanks!

**Pull request merged**